source/agora/common/TransactionManager.d

```diff
@@ -0,0 +1,283 @@
+
+        Return:
+            The input of the transaction
+
+    *******************************************************************************/
+
+    public Input newTxIn (Hash previous, uint index, Signature signature = TransactionManager.nullSignature)
+    {
+        Input txIn;
+        txIn.previous = previous;
+        txIn.index = index;
+        txIn.signature = signature;
+        return txIn;
+    }
+
+    /*******************************************************************************
+
+        New `Output` of `Transaction`
+
+        Params:
+            address =  The public key that can redeem this output
+            value = The value of this output
+
+        Return:
+            The output of the transaction
+
+    *******************************************************************************/
+
+    public Output newTxOut (PublicKey address, Amount value = 0)
+    {
+        Output txOut;
+        txOut.address = address;
+        txOut.value = value;
+        return txOut;
+    }
+
+    /*******************************************************************************
+
+        Checks whether the transaction is coinbase
+
+        Params:
+            tx = Type of `Transaction`
+
+        Return:
+            Return true if this transaction is a coinbase
+
+    *******************************************************************************/
+
+    public bool isCoinbaseTx (Transaction tx)
+    {
+        return
+            (
+                (tx.inputs.length == 1) &&
+                (tx.inputs[0].previous == TransactionManager.nullTxHash) &&
+                (tx.inputs[0].index == 0) &&
+                (tx.inputs[0].signature == TransactionManager.nullSignature)
+            );
+    }
+
+    /*******************************************************************************
+
+        Creates a new coinbase transaction
+
+        Params:
+            address = The public key that can redeem this output
+            value = The initial value
+
+        Return:
+            Return coinbase transation
+
+    *******************************************************************************/
+
+    public Transaction newCoinbaseTX (PublicKey address, Amount value = 0)
+    {
+        Transaction tx;
+        tx.inputs ~= this.newTxIn(TransactionManager.nullTxHash, 0);
+        tx.outputs ~= this.newTxOut(address, value);
+        return tx;
+    }
+
+    /*******************************************************************************
+
+        Find a output of transaction
+
+        Params:
+            hash = `Hash` of `Transaction`
+            index = Index of the `Output` in the `Transaction`
+
+        Return:
+            `Output` in the `Transaction`
+
+    *******************************************************************************/
+
+    public Output * findTxOutput (Hash hash, uint index)
+    {
+        if (hash in this.txs)
```

Comment:
This is doing a double lookup. The builtin hashmap returns a pointer to the element if it exists, or `null`.
Additionally, you can define a new variable in `if` statement, so the following code:
```d
if (auto tx = hash in this.txs)
{
    // Use `tx`
}
else
{
    // Cannot use `tx` from here, but we know it is `null`
    return null;
}
```

Works just fine.